nixos/kanidm: make db_path option writeable

TheRealGramdalf opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

Removed the read-only stipulation from `services.kanidm.serverSettings.db_path` and added a warning with necessary workarounds. This is intended as a temporary fix for #323674, which should fix #312012 by proxy. Supersedes #323676.

This solution has been partially confirmed as working by abusing environment variables to set `KANIDM_DB_PATH`. As of my recent upgrade from Kanidm 1.5.X to 1.6.X this no longer works (since the toml file takes precedence), but the underlying principle should still hold after this change. A release note was not added as this will not affect existing users, though it could be added if desired to notify users who may want to make use of the new functionality.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

TheRealGramdalf committed:
nixos/kanidm: make db_path option writeable

PatrickDaG commented on nixos/modules/services/security/kanidm.nix:
```diff
@@ -282,9 +282,19 @@ in
+              ::: {.warning
+              Please note that while it is now possible to modify this option, extra care
+              must be taken due to hardening options set on the primary kanidm daemon.
+              It is recommended to avoid putting the database in a folder with
```
Also I feel like this hint isn't really on topic as the problems with other files is more on the other files, than on the `db_path`.

I think the problem that you should mention is that when changing the database away from `/var/lib/kanidm` you will have to take extra care that that path is accessible and writable yourself.

TheRealGramdalf replied:
This is based on my experience using the env to override - it's been a while so I could well be wrong, but I believe a separate folder is necessary since due to the way systemd logic/module logic works it would become readonly.

PatrickDaG reviewed (changes requested):
I don't think this is actually a fix for #312012, as with the same options set that will still not work. The fix for that probably includes checking that no path is in both 'BindReadOnly` and `BindPath` options.

TheRealGramdalf commented:
> Also are you sure the toml file takes precedence? I'm fairly certain it's not supposed to. ENV should still override.

I took a look over it and that should be correct, but for whatever reason it wasn't working as it was previously with `systemd.services."kanidm".environment."KANIDM_DB_PATH" = "${dataDir}/db/kanidm.db";` - it somehow went back to `/var/lib/kanidm`. I verified it with `systemctl cat kanidm` as well, which gave
```
[Service]
Environment="KANIDM_DB_PATH=/persist/services/kanidm/db/kanidm.db"
```
Not sure why that's happening.

TheRealGramdalf commented:
Status update: This PR is waiting on #409310, which removes the problematic logic mentioned. This PR can then be simplified to making the option writable, and then either notifying the user/adding the databases parent directory to `ReadWritePaths`

mweinelt commented:
Have you considered a bind mount to swap out the backing store below /var/lib/kanidm? If so, why doesn't that work for you?

Leaving the module as-is simplifies maintenance, because fewer considerations need to be made about the management of the `StateDirectory=`.

TheRealGramdalf commented:
> It lets a module provision a state directory without much fuss and in a way that works for most users.

> I think that is actually a great permanent solution. It allows keeping modules lean and simple, while allowing all the needed flexibility of choosing where things get stored in the end. It has just become a standard pattern for me to mount ZFS datasets as required and systemd supports that really well:

While I'm sure it works great for you, it doesn't work for everybody. Also keep in mind that a user shouldn't have to resort to filesystem mounts to work around a self-created problem - it's quite trivial on ZFS, yes, but not everyone uses that. Not everyone is familiar with bind mounts either. On non-zfs, bind mounts can be somewhat finnicky or fail unexpectedly - this is especially bad considering Kanidm supplies authentication - if it goes down, it most likely takes down a lot of your infrastructure with it. That's bad.

For those reasons (among others I've left out for the sake of brevity), I believe this option should become read-write. The only reason it was read-only in the first place (to the best of my knowledge - it isn't actually documented or explained in the git history, it was just made that way in the initial PR adding Kanidm) was due to hardening options, in combination with the logic for adding `ReadOnlyPaths` not functioning properly. The latter is in the process of being resolved, and the removal of the the `Read*Paths` function will greatly simplify this PR. 

**In summary, please withhold reviews until #409310 is merged and I've rebased on top of it**

mweinelt commented:
>  it's quite trivial on ZFS, yes, but not everyone uses that

Bind mounts are trivial on every filesystem and don't require ZFS. 

> Not everyone is familiar with bind mounts either

The concept is simple enough. 

1. Create a directory at the location you actually want to store the data at
2. Set up a mount configuration for `/var/lib/kanidm` backed by directory created in step 1 and use the `bind` option.

```nix
  fileSystems."/var/lib/kanidm" = {
    device = "/mnt/storage/kanidm";
    options = [ "bind" ];
  };
```

TheRealGramdalf commented:
> Anyway, I'm shutting up now.

Apologies if I came across as pushy or "I'm right you're wrong", that wasn't my intention - I'm just trying to reason through it as the benefits and downsides didn't line up for me. Sorry!

> You'll find this kind of set up with various different services in nixpkgs.

AFAICT it's only prowlarr that does this, feel free to correct me if I'm wrong. I'm beginning to think that nixpkgs needs a formal RFC for how users/groups, ownership of files, and hardening options of services work, since implementations seem to vary quite a bit - some have `tmpfiles` rules conditional on the user/group remaining default, some do it unconditionally, some (few?) don't manage ownership at all, etc. etc. It would be nice to have a standard to go by as to how to decide and how to implement things - what to manage in the module and what to leave to the users' discretion.

adamcstephens commented:
I'd be concerned this is a footgun as proposed. If we want to allow changing `db_path`, then we need to ensure it gets added to BindPaths if not the default of `/var/lib/kanidm`. This increases the logic in this already non-trivial module, but seems necessary to make the proposed change.

I agree there's a lack of consistency across modules, but RFC+implementation+enforcement of some soft of standardization would be a long and painful project for anyone to undertake. I wouldn't discourage it, but that is the reality.

bjornfor commented:
AFAIK, a benefit of using bind mounts is that that paths seen by the application stay the same. So data can be moved around, without having to deal with data migration in the app.

TheRealGramdalf commented:
Rebased on master, and changed the PR accordingly. This is probably the simplest way to implement moving of the DB, leaving it up to the end user.

Alternatively, the parent folder of the DB could be added automatically either as a `ReadWritePath`/as the `StateDirectory`, however this requires additional logic - for some reason I thought `lib.parentDirectory` or the like existed which would have made this simple, but as it is it would require an additional function to determine the parent directory - the `db_path` is specified as a path to the file, not the directory containing it.

Does a function like that make sense, or is there another, better way of doing it?

TheRealGramdalf commented:
I went ahead and added the function, as there doesn't seem to be a better option. A few notes on the implementation:

- I could have used `lib.path.subpath.*`, but it would require splitting it into a subpath with `lib.path.splitRoot`. If `lib.path.subpath` is preferred I can make that happen
- I split the string trickery into a separate function to hopefully make the service declaration more readable compared to doing it inline
- I thought more about using `BindPaths` as mentioned above, setting the option in the toml file to `/var/lib/kanidm/kanidm.db` unconditionally, so that `BindPaths` could mount the user specified directory on top - but realized that in doing so, the config file in `/etc/` would not point to the correct location when the unit is stopped - this matters because `kanidmd domain upgrade-check` is run outside of the systemd unit, and accesses the database using the path in the config file

Let me know what your thoughts are!

emilylange commented:
This feels a bit like as if folks are talking past each other.

@TheRealGramdalf, the suggestion isn't to somehow implement bind-mounts in the `nixos/kanidm` module.
You, as an end-user wanting this, can and should use bind-mounts in your own local config along where you call and configure `services.kanidm`, not in the module itself.

This PR adds unwarranted complexity and fragileness, as you happen to acknowledge yourself in 2cb9bcf41cc83235711a5b8a3e9fd7e550436088:

> ::: {.warning}
> Please note that while it is possible to modify this option, extra care
> must be taken due to hardening options set on the primary kanidm daemon.
> If changed from the default, the folder containing the kanidm database
> must be added to `systemd.services."kanidm".serviceConfig.ReadWritePaths`
> :::

Using bind-mounts on the other hand is already possible, universal, transparent, very simple, extremely robust thanks to how systemd already adds and handles the necessary dependencies automatically in the background and unlikely to break due to some future modification made to this module or package.
